includes/widget-css-classes.class.php

```diff
@@ -152,114 +349,233 @@ public static function add_widget_classes( $params ) {
 
-			if ( 1 === (int) $widget_css_classes['type'] || 3 === (int) $widget_css_classes['type'] ) {
+			$custom_classes = explode( ' ', (string) $widget_opt[ $widget_num ]['classes'] );
+			$defined_classes = WCSSC_Lib::get_settings( 'defined_classes' );
+
+			// Order classes by predefined classes order and append the other (custom) classes.
+			if ( ! empty( $defined_classes ) ) {
+				// Order classes selection by predefined classes order and append the other (custom) classes.
+				$custom_classes = array_filter( array_unique( array_merge( array_intersect( $defined_classes, $custom_classes ), $custom_classes ) ) );
+			}
+
+			/**
+			 * Modify the list of custom CSS classes.
+			 * Can also be used for ordering etc.
+			 *
+			 * @since  1.5.0
+			 * @param  array      $custom_classes
+			 * @param  string     $widget_id
+			 * @param  int        $widget_number
+			 * @param  array      $widget_opt
+			 * @param  WP_Widget  $widget_obj
+			 * @return array
+			 */
+			$custom_classes = (array) apply_filters( 'widget_css_classes_custom', $custom_classes, $widget_id, $widget_number, $widget_opt, $widget_obj );
+
+			$type = WCSSC_Lib::get_settings( 'type' );
+
+			if ( 1 === (int) $type || 3 === (int) $type ) {
 				// Add all classes
-				$params[0]['before_widget'] = preg_replace( '/class="/', "class=\"{$widget_opt[ $widget_num ]['classes']} ", $params[0]['before_widget'], 1 );
-			} elseif ( 2 === (int) $widget_css_classes['type'] ) {
+				$classes = array_merge( $classes, $custom_classes );
+			} elseif ( 2 === (int) $type ) {
 				// Only add predefined classes
-				$predefined_classes = explode( ';', $widget_css_classes['defined_classes'] );
-				$classes = explode( ' ', $widget_opt[ $widget_num ]['classes'] );
-				foreach ( $classes as $key => $value ) {
-					if ( in_array( $value, $predefined_classes, true ) ) {
-						$value = esc_attr( $value );
-						$params[0]['before_widget'] = preg_replace( '/class="/', "class=\"{$value} ", $params[0]['before_widget'], 1 );
+				foreach ( $custom_classes as $key => $value ) {
+					if ( in_array( $value, $defined_classes, true ) ) {
+						$classes[] = $value;
 					}
 				}
 			}
-		}
+		} // End if().
 
-		// Add id
-		if ( 1 === (int) $widget_css_classes['show_id'] ) {
-			if ( isset( $widget_opt[ $widget_num ]['ids'] ) && ! empty( $widget_opt[ $widget_num ]['ids'] ) )
-				$params[0]['before_widget'] = preg_replace( '/id="[^"]*/', "id=\"{$widget_opt[ $widget_num ]['ids']}", $params[0]['before_widget'], 1 );
-		}
+		// Add first, last, even, and odd classes.
+		if ( WCSSC_Lib::get_settings( 'show_number' ) ||
+		     WCSSC_Lib::get_settings( 'show_location' ) ||
+		     WCSSC_Lib::get_settings( 'show_evenodd' )
+		) {
 
-		// Add first, last, even, and odd classes
-		if ( 1 === (int) $widget_css_classes['show_number'] || 1 === (int) $widget_css_classes['show_location'] || 1 === (int) $widget_css_classes['show_evenodd'] ) {
 			if ( ! $widget_number ) {
 				$widget_number = array();

```

Comment:
@cleverness 
At the start of this method `$widget_number` is declared as a global variable. Is there a specific reason to make it a global variable? It doesn't look like it's supposed to be modified externally so maybe it's better to declare it as a static variable inside this method.
Let me know!